### PR TITLE
Remove unused props from default templates and save requests

### LINKED ISSUE #12262: Avoid persistently storing certain properties
## Feature Description

This is a follow-up to #12131.

1. Do not persistently store the `basedOn` property. Seems useless to keep, since this is only used when duplicating elements.

2. Do not persistently store global properties things like `current`, `selection`, and `story`  
  Update: actually, I don't think we store these in the DB. Just need to remove them from templates.

### PR DESCRIPTION
**Pull request by ayushnirwal**
## Summary

- Removes `current`, `selection` and `story` props from default template.
- Updates `getStoryPropsToSave` to ignore `basedOn` property in elements.
- Updates type `StorySaveData`.
- Adds migration for removing `basedOn` property.

## To-do
- [x] Use selected element list to derive offset for staggering effect.
  - [x] Update related tests.
- [x] remove `basedOn` from the types.
- [x] change`getStroryPropsToSave` to not delete `basedOn` attribute.

## Testing Instructions

- [ ] This is a non-user-facing change and requires no QA

This PR can be tested by following these steps:

1.

## Checklist

- [ ] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [ ] I have tested this code to the best of my abilities
- [ ] I have verified accessibility to the best of my abilities ([docs](https://github.com/googleforcreators/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [ ] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [ ] This code is covered by automated tests (unit, integration, and/or e2e) to verify it works as intended ([docs](https://github.com/googleforcreators/web-stories-wp/tree/main/docs#testing))
- [ ] I have added documentation where necessary
- [ ] I have added a matching `Type: XYZ` label to the PR

---

Fixes #12262

### CODE REVIEW

**Comment by swissspidy** [comment]
> * Removes `current`, `selection` and `story` props from default template.

Let's also remove them in `packages/story-editor/src/components/devTools/devTools.js` when enabling template mode (in `prepareTemplate()`)

> * Updates `getStroryPropsToSave` to ignore `basedOn` property in elements.
> * Updates type `StorySaveData`.
> * Adds migration for removing `basedOn` property.

With that in place we should properly refactor `duplicateElement` to not store `basedOn` persistently. Because if we remove it there, then we don't even need to change `getStroryPropsToSave`; the migration would be enough.

Thoughts?

**Comment by ayushnirwal** [comment]
> With that in place we should properly refactor duplicateElement to not store basedOn persistently. Because if we remove it there, then we don't even need to change getStroryPropsToSave; the migration would be enough.

Is there no use case of saving `basedOn` in the provider state?

**Comment by swissspidy** [comment]
That's what I don't know :-) It's been a while since this was added, so I am lacking the context.

**Comment by ayushnirwal** [comment]
> Is there no use case of saving basedOn in the provider state?

Found one use case here.
https://github.com/GoogleForCreators/web-stories-wp/blob/80d319272c6d992396038284f126bda8f7a43bcb/packages/elements/src/utils/duplicateElement.ts#L76-L88

These lines of code are responsible for calculating the positional offset based on the number of duplicates how many duplicates an element already has.

**Comment by swissspidy** [comment]
> These lines of code are responsible for calculating the positional offset based on the number of duplicates how many duplicates an element already has.

So the more duplicates an element has, the bigger is the offset 🤔 Interesting way to achieve this effect:

https://user-images.githubusercontent.com/841956/234823212-a8c8b742-7b64-4fbf-b0f0-df9cf7d14b2e.mov

Especially since it doesn't really account for the positions of all the existing elements, leading to unexpected offsets like this:

https://user-images.githubusercontent.com/841956/234823427-3cc950e4-d369-4849-8870-ba497ea78407.mov

I think in the end it comes all back to #1206, where we originally envisioned nice staggering behavior when inserting multiple elements, leading to solutions like this and also #4218

Can we maybe find a simpler staggering behavior for duplicating elements without having to use `basedOn`?

**Comment by ayushnirwal** [comment]
@swissspidy Since the pasted elements becomes selected, offset can be calculated off of that.

File: packages/elements/src/utils/duplicateElement.ts [hunk 1]
```diff
@@ -33,7 +33,7 @@ import createNewElement from './createNewElement';
  * @param originY Original Y.
  * @return Coordinates.
  */
-function getOffsetCoordinates(originX: number, originY: number) {
+export function getOffsetCoordinates(originX: number, originY: number) {
```
Comment (swissspidy) [inline packages/elements/src/utils/duplicateElement.ts]: Now that this is exported, should we move it to its own `getOffsetCoordinates.ts` file for visibility & maintenance?

File: packages/story-editor/src/types/story.ts [hunk 2]
```diff
@@ -36,6 +37,12 @@ interface FeaturedMedia {
   isExternal: boolean;
 }
 
+export type ElementSaveData = Element;
+
+export interface PageSaveData extends Omit<Page, 'elements'> {
+  elements: ElementSaveData[];
+}
+
```
Comment (swissspidy) [inline packages/story-editor/src/types/story.ts]: Should we remove this again now that `ElementSaveData` is not different anymore?

File: packages/story-editor/src/utils/copyPaste.tsx [hunk 3]
```diff
@@ -76,15 +78,21 @@ export function processPastedElements(
           elements,
           animations,
         }: { elements: Element[]; animations: StoryAnimation[] },
-        payloadElement: Element
+        payloadElement: Element,
+        ind: number
```
Comment (swissspidy) [inline packages/story-editor/src/utils/copyPaste.tsx]: Nit: `ind` -> `index` for readability

**Review by swissspidy — APPROVED** [review]
This works really well! Kudos for making this simplification happen.

Left just a few tiny comments, otherwise LGTM

File: packages/elements/src/utils/index.ts [hunk 4]
```diff
@@ -20,5 +20,8 @@ export { default as duplicatePage } from './duplicatePage';
 export { default as isElementBelowLimit } from './isElementBelowLimit';
 export { default as getTransformFlip } from './getTransformFlip';
 export { default as getLayerName } from './getLayerName';
-export { default as duplicateElement } from './duplicateElement';
+export {
+  default as duplicateElement,
+  getOffsetCoordinates,
```
Comment (swissspidy) [inline packages/elements/src/utils/index.ts]: This is now in its own separate file

File: packages/story-editor/src/app/story/utils/getAllProducts.ts [hunk 5]
```diff
@@ -20,15 +20,19 @@
...
+/**
+ * Internal dependencies
+ */
+import type { PageSaveData } from '../../../types';
+
 function isProduct(e: Element): e is ProductElement {
   return 'product' in e;
 }
-function getAllProducts(pages: Page[]): ProductData[] {
+function getAllProducts(pages: PageSaveData[]): ProductData[] {
```
Comment (swissspidy) [inline packages/story-editor/src/app/story/utils/getAllProducts.ts]: This now needs to be undone as `PageSaveData` doesn't exist anymore

**Pull request merged**